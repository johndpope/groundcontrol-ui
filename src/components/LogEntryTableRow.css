.ui.table .LogEntryTableRow td {
  word-break: break-all;
  white-space: pre-wrap;
  font: var(--mono-font);
  padding-top: .25em;
  padding-bottom: .25em;
  border-top: none;
}

.ui.table .LogEntryTableRow td code {
  font-size: 1em;
}
