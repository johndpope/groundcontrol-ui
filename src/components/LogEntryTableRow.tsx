// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import graphql from "babel-plugin-relay/macro";
import React, { Fragment } from "react";
import { Responsive, Table } from "semantic-ui-react";

import Moment from "react-moment";
import { createFragmentContainer } from "react-relay";

import { LogEntryTableRow_item } from "./__generated__/LogEntryTableRow_item.graphql";
import { LogEntryTableRow_prevItem } from "./__generated__/LogEntryTableRow_prevItem.graphql";

import LogEntryMessage, {
  IProps as ILogEntryMessageProps
} from "./LogEntryMessage";

import "./LogEntryTableRow.css";

const dateFormat = "LTS";

export interface IProps {
  item: LogEntryTableRow_item;
  prevItem: LogEntryTableRow_prevItem | null;
  onClickSourceFile: (values: ILogEntryMessageProps) => any;
}

type Owner = LogEntryTableRow_item["owner"];

type LongStringer = Owner & {
  longString: string;
};

function isLongStringer(owner: Owner): owner is LongStringer {
  return !!owner && !!owner.longString;
}

type Stringer = Owner & {
  string: string;
};

function isStringer(owner: Owner): owner is Stringer {
  return !!owner && !!owner.string;
}

export function LogEntryTableRow(props: IProps) {
  const {
    item,
    item: { level },
    onClickSourceFile
  } = props;

  const showMeta = shouldShowMeta(props);

  return (
    <Table.Row className="LogEntryTableRow" verticalAlign="top">
      <Responsive
        as={Table.Cell}
        className="LogEntryTableRowCreatedAt"
        minWidth={992}
        collapsing={true}
      >
        {showMeta && <CreatedAt {...props} />}
      </Responsive>
      <Table.Cell collapsing={true}>
        {showMeta && <Namespace {...props} />}
      </Table.Cell>
      <Table.Cell
        className="LogEntryTableRowLevel"
        warning={level === "WARNING"}
        error={level === "ERROR"}
        collapsing={true}
      >
        {showMeta && level}
      </Table.Cell>
      <Table.Cell>
        <LogEntryMessage item={item} onClickSourceFile={onClickSourceFile} />
      </Table.Cell>
    </Table.Row>
  );
}

const CreatedAt = ({ item: { createdAt } }: IProps) => (
  <Moment format={dateFormat}>{createdAt}</Moment>
);

export const Namespace = ({ item: { owner } }: IProps) => {
  if (!owner) {
    return null;
  }

  if (isLongStringer(owner)) {
    return <Fragment>{owner.longString}</Fragment>;
  }

  if (isStringer(owner)) {
    return <Fragment>{owner.string}</Fragment>;
  }

  return <Fragment>owner.id</Fragment>;
};

function shouldShowMeta(props: IProps) {
  const {
    item: { createdAt, owner, level },
    prevItem
  } = props;

  if (!prevItem) {
    return true;
  }

  if (!owner !== !prevItem.owner) {
    return true;
  }

  if (owner && prevItem.owner && owner.id !== prevItem.owner.id) {
    return true;
  }

  return prevItem.level !== level || prevItem.createdAt !== createdAt;
}

export default createFragmentContainer(
  LogEntryTableRow,
  graphql`
    fragment LogEntryTableRow_item on LogEntry {
      createdAt
      level
      ...LogEntryMessage_item
      owner {
        id
        ... on LongStringer {
          longString
        }
        ... on Stringer {
          string
        }
      }
    }
    fragment LogEntryTableRow_prevItem on LogEntry {
      createdAt
      level
      owner {
        id
      }
    }
  `
);
