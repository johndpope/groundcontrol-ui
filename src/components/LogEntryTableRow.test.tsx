// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallow } from "enzyme";
import React from "react";

import { LogEntryTableRow_item } from "./__generated__/LogEntryTableRow_item.graphql";
import { LogEntryTableRow_prevItem } from "./__generated__/LogEntryTableRow_prevItem.graphql";

import { mockQueryPropAttrs } from "../testing/relay";

import { LogEntryTableRow, Namespace } from "./LogEntryTableRow";

const other: LogEntryTableRow_item = {
  ...mockQueryPropAttrs(),
  createdAt: "createdAt",
  level: "INFO",
  owner: {
    id: "ownerId"
  }
};

const stringer: LogEntryTableRow_item = {
  ...mockQueryPropAttrs(),
  createdAt: "createdAt",
  level: "INFO",
  owner: {
    id: "ownerId",
    string: "string"
  }
};

const longStringer: LogEntryTableRow_item = {
  ...mockQueryPropAttrs(),
  createdAt: "createdAt",
  level: "INFO",
  owner: {
    id: "ownerId",
    longString: "longString"
  }
};

const prevItem: LogEntryTableRow_prevItem = {
  ...mockQueryPropAttrs(),
  createdAt: "createdAt",
  level: "INFO",
  owner: {
    id: "ownerId"
  }
};

const otherProps = {
  item: other,
  onClickSourceFile: jest.fn(),
  prevItem: null
};

const stringerProps = {
  item: stringer,
  onClickSourceFile: jest.fn(),
  prevItem: null
};

const longStringerProps = {
  item: longStringer,
  onClickSourceFile: jest.fn(),
  prevItem: null
};

const similarProps = {
  item: other,
  onClickSourceFile: jest.fn(),
  prevItem
};

describe("<Namespace />", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<Namespace {...otherProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly when the owner is a stringer", () => {
    const wrapper = shallow(<Namespace {...stringerProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly when the owner is a longStringer", () => {
    const wrapper = shallow(<Namespace {...longStringerProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("<LogEntryTableRow />", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<LogEntryTableRow {...otherProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("sets the warning attribute of the level cell to true when the level is warning", () => {
    const wrapper = shallow(
      <LogEntryTableRow
        {...{ ...otherProps, item: { ...other, level: "WARNING" } }}
      />
    );
    expect(wrapper.find("[warning=true]")).toHaveLength(1);
  });

  it("sets the error attribute of the status cell to true when the level is error", () => {
    const wrapper = shallow(
      <LogEntryTableRow
        {...{ ...otherProps, item: { ...other, level: "ERROR" } }}
      />
    );
    expect(wrapper.find("[error=true]")).toHaveLength(1);
  });

  it("doesn't display meta if the previous entry is similar", () => {
    const wrapper = shallow(<LogEntryTableRow {...similarProps} />);
    expect(wrapper.find("Namespacej")).toHaveLength(0);
  });

  it("displays meta if the previous entry's createdAt is different", () => {
    const wrapper = shallow(
      <LogEntryTableRow
        {...{ ...otherProps, prevItem: { ...prevItem, createdAt: "new" } }}
      />
    );
    expect(wrapper.find("Namespace")).toHaveLength(1);
  });

  it("displays meta if the previous entry's level is different", () => {
    const wrapper = shallow(
      <LogEntryTableRow
        {...{ ...otherProps, prevItem: { ...prevItem, level: "DEBUG" } }}
      />
    );
    expect(wrapper.find("Namespace")).toHaveLength(1);
  });

  it("displays meta if the previous entry's owner is different", () => {
    const wrapper = shallow(
      <LogEntryTableRow
        {...{ ...otherProps, prevItem: { ...prevItem, owner: null } }}
      />
    );
    expect(wrapper.find("Namespace")).toHaveLength(1);
  });
  it("displays meta if the previous entry's owner id is different", () => {
    const wrapper = shallow(
      <LogEntryTableRow
        {...{ ...otherProps, prevItem: { ...prevItem, owner: { id: "new" } } }}
      />
    );
    expect(wrapper.find("Namespace")).toHaveLength(1);
  });
});
