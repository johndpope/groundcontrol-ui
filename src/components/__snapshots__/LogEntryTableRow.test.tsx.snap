// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LogEntryTableRow /> renders correctly 1`] = `
<TableRow
  as="tr"
  cellAs="td"
  className="LogEntryTableRow"
  verticalAlign="top"
>
  <Responsive
    as={[Function]}
    className="LogEntryTableRowCreatedAt"
    collapsing={true}
    getWidth={[Function]}
    minWidth={992}
  >
    <CreatedAt
      item={
        Object {
          " $fragmentRefs": null,
          " $refType": null,
          "createdAt": "createdAt",
          "level": "INFO",
          "owner": Object {
            "id": "ownerId",
          },
        }
      }
      onClickSourceFile={[MockFunction]}
      prevItem={null}
    />
  </Responsive>
  <TableCell
    as="td"
    collapsing={true}
  >
    <Namespace
      item={
        Object {
          " $fragmentRefs": null,
          " $refType": null,
          "createdAt": "createdAt",
          "level": "INFO",
          "owner": Object {
            "id": "ownerId",
          },
        }
      }
      onClickSourceFile={[MockFunction]}
      prevItem={null}
    />
  </TableCell>
  <TableCell
    as="td"
    className="LogEntryTableRowLevel"
    collapsing={true}
    error={false}
    warning={false}
  >
    INFO
  </TableCell>
  <TableCell
    as="td"
  >
    <ForwardRef(Relay(LogEntryMessage))
      item={
        Object {
          " $fragmentRefs": null,
          " $refType": null,
          "createdAt": "createdAt",
          "level": "INFO",
          "owner": Object {
            "id": "ownerId",
          },
        }
      }
      onClickSourceFile={[MockFunction]}
    />
  </TableCell>
</TableRow>
`;

exports[`<Namespace /> renders correctly 1`] = `
<Fragment>
  owner.id
</Fragment>
`;

exports[`<Namespace /> renders correctly when the owner is a longStringer 1`] = `
<Fragment>
  longString
</Fragment>
`;

exports[`<Namespace /> renders correctly when the owner is a stringer 1`] = `
<Fragment>
  string
</Fragment>
`;
